### DISCUSSION TITLE
Anyone using SolarEdge API?

### DISCUSSION THREAD

**Discussion by youzer-name** [jasonacox/Powerwall-Dashboard#162] [Show and tell]
I'm going to be adding additional solar to my roof soon and it will use a SolarEdge inverter. I am hoping to pull data from the SolarEdge API so I can continue to calculate costs and savings from my system. 

The existing solar is a Tesla/Solar City PPA (so I pay for the solar it produces) and I will own the new panels. The two inverter outputs are going to be combined in a subpanel before they are fed to the Powerwall gateway, so Powerwall Dashboard is only going to show me total production. I'm going to need to log the SolarEdge production and subtract it from the total to get the PPA production.

If anyone is logging data from SolarEdge, what project are you using? I see multiple projects here on github, so it would be helpful to know what is working for others here.

**Comment by youzer-name** [comment]
Following up on my own post, and moving it to the show & tell category.

After trying out a few options for how to pull data from the SolarEdge API, I ended up using a python webscraper that I found online that pulls data from their monitoring website. The website provides optimizer-level data where the API does not. The downside to using the website is that you are limited to pulling at most the last 7 days of data, so I wasn't able to go back and pull all the data since the system was installed. I might try to manually backfill the total output data for the dates before I started scraping the data from the website.

These are some of the panels I've built from the SolarEdge data:

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/57046429/68a5feb1-2dc3-44bc-9169-2865210ac630)

These show the two roof planes (front & rear) with the individual panel production for the last 24 hours. The charts on the left show the production stacked, so the scale is the total output of the system. The charts on the right show the production without stacking, so it is easier to compare one panel vs. another. You can see that on both roof planes I have two panels that get throttled by the optimizers for part of the day due to shading.

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/57046429/e4aac4f6-2427-42ed-bd0c-f296d1ee3487)

The panel on the left shows the full roof layout with the background shading indicating the relative output (using color mode: background and color scheme 'blues by value'). The panels all the way to the right shows the same thing, but shows the front and back roof in separate panels which results in slightly different relative shading. The panels in the center show the grand total production and the front/rear total production.

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/57046429/fc57c7a7-9e9c-415c-8266-a656ad7857de)

These panels show the solar production measured by the Powerwall gateway (in orange), the production by my SolarCity/Tesla array (yellow) and my SolarEdge array (green) with the Tesla and SolarEdge stacked. Since I am on non-plus Powerwall Gateway 2, the Tesla solar data is coming from the Solar-Only history tool and like the SolarEdge data, has 15 minute resolution. 

The "extra" shown in the stat panel is the production measured by the Powerwall that is above (or below) the total of the two systems. It's usually a positive number, but I've seen it negative a few times.

Also, unrelated to SolarEdge, I wanted to show off my Grid Usage panel. I believe the default one for the project dashboard looks like this:

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/57046429/7f5624a3-b16f-4f6d-ae8f-3101c1289ea8)


I changed the colors on mine, but also added a single point on each day (in yellow) that shows the net grid usage for that day. I think the net daily usage adds a lot to this panel. Mine now looks like this:

![image](https://github.com/jasonacox/Powerwall-Dashboard/assets/57046429/8bbd615a-a36f-425a-b487-449d787937bf)

**Reply by jasonacox** [reply]
Wow!  This is awesome @youzer-name !  Please do share your 'python webscraper` tricks! 😁 

> I wanted to show off my Grid Usage panel. 

That's beautiful!

**Comment by youzer-name** [comment]
@jasonacox 

I am barely proficient in Python, so I can't claim credit for the webscraper. I think I originally found it here:  https://gist.github.com/mihailescu2m/27d5dd22656a4d64a0c755da99b7b162. I have some idea of how it works, but I definitely couldn't have created this myself.

I run it from a cron job every 15 minutes during daytime hours. Since my inverter goes offline fairly often, I also run the 'weekly' version once a day so it will fill in any missing data for the past week. To pull weekly data you replace "DAILY_DATA" with "WEEKLY_DATA" on lines 27, 33, and 36 of the script.

I also modified the script to add logging, as I was having an issue where it was working manually but hanging up when run from the cron and I needed the logging to see what was going on. I think the issue was due to cron job saving the pickle file in a different directory than where it got saved when I ran it manually. I deleted the pickle file that was in the user home directory and it started to run successfully on schedule again.

**Reply by jasonacox** [reply]
Thanks @youzer-name !
